File: pose_view_android/android/src/main/kotlin/com/relive/poseview/camera/NativeCameraViewFactory.kt
package com.relive.poseview.camera

import android.app.Activity
import android.content.Context
import androidx.lifecycle.LifecycleOwner
import com.relive.poseview.ViewController
import com.relive.poseview.MainViewModel
import com.relive.poseview.PoseDataStreamHandler
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.StandardMessageCodec
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory

class NativeCameraViewFactory(
    private val activity: Activity,
    private val owner: LifecycleOwner,
    private val cameraController: ViewController,
    private val poseDataStreamHandler: PoseDataStreamHandler,
    private val messenger: BinaryMessenger
) : PlatformViewFactory(StandardMessageCodec.INSTANCE) {

    override fun create(context: Context, viewId: Int, args: Any?): PlatformView {
        val creationParams = args as Map<String?, Any?>?

        // Create ViewModel from options
        val options = creationParams?.get("options") as Map<String, Any>
        val viewModel = MainViewModel(
            options["model"] as Int,
            options["delegate"] as Int,
            (options["minPoseDetectionConfidence"] as Double).toFloat(),
            (options["minPoseTrackingConfidence"] as Double).toFloat(),
            (options["minPosePresenceConfidence"] as Double).toFloat(),
        )

        return NativeCameraView(context, viewModel, activity, owner, cameraController, poseDataStreamHandler, messenger, viewId)
    }
}

File: pose_view_ios/ios/Classes/Camera/NativeCameraViewFactory.swift
//
//  FLNativeView.swift
//  Runner
//
//  Created by ios on 06/12/2023.
//

import Flutter
import UIKit

@available(iOS 13.0, *)
class NativeCameraViewFactory: NSObject, FlutterPlatformViewFactory {
  private var registrar: FlutterPluginRegistrar
  private var poseDataStreamHandler: PoseDataStreamHandler

  init(registrar: FlutterPluginRegistrar, poseDataStreamHandler: PoseDataStreamHandler) {
    self.registrar = registrar
    self.poseDataStreamHandler = poseDataStreamHandler
    super.init()
  }

  func create(
    withFrame frame: CGRect,
    viewIdentifier viewId: Int64,
    arguments args: Any?
  ) -> FlutterPlatformView {
    let creationParams = args as! [String: Any]
      
    let options = creationParams["options"] as! [String: Any]

    let modelValue = options["model"] as! Int
    let model = Model(rawValue: modelValue)!

    let delegateValue = options["delegate"] as! Int
    let delegate: PoseLandmarkerDelegate = {
        switch delegateValue {
        case 0:
            return .CPU
        case 1:
            return .GPU
        default:
            fatalError("Invalid delegate value")
        }
    }()

    let inferenceConfig = InferenceConfigurationManager(
      model: model,
      delegate: delegate,
      minPoseDetectionConfidence: options["minPoseDetectionConfidence"] as! Float,
      minPosePresenceConfidence: options["minPosePresenceConfidence"] as! Float,
      minTrackingConfidence: options["minPoseTrackingConfidence"] as! Float
    )

    return NativeCameraView(
      frame: frame,
      viewIdentifier: viewId,
      arguments: args,
      binaryMessenger: registrar.messenger(),
      registrar: registrar,
      poseDataStreamHandler: poseDataStreamHandler,
      inferenceConfig: inferenceConfig
    )
  }

  /// Implementing this method is only necessary when the `arguments` in `createWithFrame` is not `nil`.
  public func createArgsCodec() -> FlutterMessageCodec & NSObjectProtocol {
    return FlutterStandardMessageCodec.sharedInstance()
  }
}
